cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DatasetModulesDeployer.java

```diff
@@ -64,58 +58,39 @@
-                           cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
-    File unpackDir = DirUtils.createTempDir(tmpDir);
-
-    try (ProgramClassLoader classLoader = getClassLoader(cConf, jarLocation, unpackDir)) {
-      for (Map.Entry<String, String> moduleEntry : modules.entrySet()) {
-        // note: using app class loader to load module class
-        @SuppressWarnings("unchecked")
-        Class<Dataset> clazz = (Class<Dataset>) classLoader.loadClass(moduleEntry.getValue());
-        String moduleName = moduleEntry.getKey();
-        try {
-          // note: we can deploy module or create module from Dataset class
-          // note: it seems dangerous to instantiate dataset module here, but this will be fine when we move deploy into
-          //       isolated user's environment (e.g. separate yarn container)
-          DatasetModuleId moduleId = namespaceId.datasetModule(moduleName);
-          if (DatasetModule.class.isAssignableFrom(clazz)) {
-            LOG.info("Adding module: {}", clazz.getName());
-            datasetFramework.addModule(moduleId.toId(), (DatasetModule) clazz.newInstance(), jarLocation);
-          } else if (Dataset.class.isAssignableFrom(clazz)) {
-            if (!systemDatasetFramework.hasSystemType(clazz.getName())) {
-              // checking if type is in already or force upgrade is allowed
-              DatasetTypeId typeId = namespaceId.datasetType(clazz.getName());
-              if (!datasetFramework.hasType(typeId.toId()) || allowDatasetUncheckedUpgrade) {
-                LOG.info("Adding module: {}", clazz.getName());
-                datasetFramework.addModule(moduleId.toId(), new SingleTypeModule(clazz), jarLocation);
-              }
+  void deployModules(NamespaceId namespaceId, Map<String, String> modules,
+                     Location jarLocation, ClassLoader artifactClassLoader) throws Exception {
+    for (Map.Entry<String, String> moduleEntry : modules.entrySet()) {
+      // note: using app class loader to load module class
+      @SuppressWarnings("unchecked")
+      Class<Dataset> clazz = (Class<Dataset>) artifactClassLoader.loadClass(moduleEntry.getValue());
+      String moduleName = moduleEntry.getKey();
+      try {
+        // note: we can deploy module or create module from Dataset class
+        // note: it seems dangerous to instantiate dataset module here, but this will be fine when we move deploy into
+        //       isolated user's environment (e.g. separate yarn container)
+        DatasetModuleId moduleId = namespaceId.datasetModule(moduleName);
+        if (DatasetModule.class.isAssignableFrom(clazz)) {
+          LOG.info("Adding module: {}", clazz.getName());
+          datasetFramework.addModule(moduleId.toId(), (DatasetModule) clazz.newInstance(), jarLocation);
+        } else if (Dataset.class.isAssignableFrom(clazz)) {
+          if (!systemDatasetFramework.hasSystemType(clazz.getName())) {
```

Comment:
This is actually the current logic. I didn't pay too much attention of it as that's not the intention of this refactoring and I was worry about changing it may break some assumptions on the DS service side. From the commit it was unclear why the check is needed.